### PR TITLE
Issue 60 stack structure

### PR DESCRIPTION
**Pull request by Liquidsoul**
## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines), and reviewed the [contributing guidelines](https://github.com/jessesquires/HowToContribute).

## What's in this pull request?
- change stack structure so that both background and main contexts are connected to the persistent store coordinator
- sync mechanism between contexts now use `mergeChangesFromContextDidSaveNotification(_:)`

## What's not in this pull request?
- child contexts creation with background context as parent.
`childContext` use the main context as parent. This may not be something we want with this new structure (at least not all the time).
- some of the tests may be simplified or changed to reflect the new stack structure and update mechanism.

### CODE REVIEW

File: Source/CoreDataStack.swift [hunk 1]
```diff
@@ -26,11 +26,11 @@ import Foundation
 
  It is composed of a main context and a background context.
  These two contexts operate on the main queue and a private background queue, respectively.
- The background context is the root level context, which is connected to the persistent store coordinator.
- The main context is a child of the background context.
+ Both are connected to the persistent store coordinator.
 
- Data between these two primary contexts and child contexts is kept in sync.
- Changes to a context are propagated to its parent context and eventually the persistent store when saving.
+ Data between these two primary contexts is kept in sync using `NSManagedObjectContextDidSaveNotification` and `mergeChangesFromContextDidSaveNotification(_:)`.
```
Comment (jessesquires) [inline Source/CoreDataStack.swift]: let's change this to:

> Data between the main and background contexts is perpetually kept in sync.

**Comment by jessesquires** [comment]
Whoa @Liquidsoul ! This is awesome! 👍 😄 Great work. 

Only one small comment to address, then I'll merge. Thanks so much for working on this!

I'll be releasing this with `4.0.0` soon.

**Comment by jessesquires** [comment]
> **What's not in this pull request?**
> - child contexts creation with background context as parent. childContext use the main context as parent. This may not be something we want with this new structure (at least not all the time).

Yep, agreed. I created #73 to follow up here. 😄 

> - some of the tests may be simplified or changed to reflect the new stack structure and update mechanism.

I think what you've done with the tests is great! 👍 Is there something specific you think we should change?

**Comment by Liquidsoul** [comment]
Thanks @jessesquires ! 😊 

I've amended my last commit to change the documentation and rebased the branch.

> I think what you've done with the tests is great! 👍 Is there something specific you think we should change?

I was wondering if the sibling context _updates_ checks in `SaveTests.swift` could be removed.
We test that saving the _background_ and the _main_ context will [merge the changes into the other](https://github.com/Liquidsoul/JSQCoreDataKit/blob/issue_60_stack_structure/Tests/SaveTests.swift#L206-L240), it may not be necessary to do [this kind of check](https://github.com/Liquidsoul/JSQCoreDataKit/blob/issue_60_stack_structure/Tests/SaveTests.swift#L57-L61) in all the tests.
Maybe this could reduce the size of some of those tests without reducing the coverage.

**Comment by jessesquires** [comment]
@Liquidsoul ahhh, I see. I actually like this. Feels very thorough. 😄

**Pull request merged**